### PR TITLE
depth estimation Example

### PR DESCRIPTION
**Pull request by victor369basu**
This is my contribution to the Keras-io computer vision example. An approach towards Depth Estimation with CNN.

### CODE REVIEW

**Comment by fchollet** [comment]
Thanks for the PR. Please only include the Python file in the PR until it has been approved (since the other files will need to be regenerated at the end).

Do you have a link to a Colab to preview the example?

**Comment by victor369basu** [comment]
I have updated with only the python file as your suggested.
link to the Colab notebook - [depth estimation](https://colab.research.google.com/drive/1zISt4f_vMpVZDuprxnboaPZbq8fYlPls?usp=sharing)

File: examples/vision/depth_estimation.py [hunk 1]
```diff
@@ -0,0 +1,507 @@
...
+"""
+## Scopes of Improvement
+
+1. From the research papers that I read while I was working on this topic the encode part
+of the unet was replaced with DenseNet, ResNet or other pre-trained model, which could be
+applied for better model predictions.
+
+2. Loss functions plays an immense role in solving this problem, and different paper
+explained different ways of developing the loss function out of which SSIM was common in
+all. so playing with the loss fuctions gives a huge scope of improvement in this case.
```
Comment (s-mrb) [inline examples/vision/depth_estimation.py:506]: comma before `so` instead of fullstop, and check spelling of functions

File: examples/vision/depth_estimation.py [hunk 2]
```diff
@@ -0,0 +1,507 @@
...
+    ax[i, 2].imshow((pred[i].squeeze()), cmap=cmap)
+
+"""
+## Scopes of Improvement
+
+1. From the research papers that I read while I was working on this topic the encode part
+of the unet was replaced with DenseNet, ResNet or other pre-trained model, which could be
+applied for better model predictions.
+
+2. Loss functions plays an immense role in solving this problem, and different paper
```
Comment (s-mrb) [inline examples/vision/depth_estimation.py:504]: > Loss functions `play` an ...

File: examples/vision/depth_estimation.py [hunk 3]
```diff
@@ -0,0 +1,507 @@
+"""
+Title: An approach towards Depth Estimation with CNN
+Author: [Victor Basu](https://www.linkedin.com/in/victor-basu-520958147)
+Date created: 2021/08/09
+Last modified: 2021/08/10
+Description: Implement an depth estimation model with CNN.
```
Comment (s-mrb) [inline examples/vision/depth_estimation.py:6]: please use `a depth` instead of `an depth`

File: examples/vision/depth_estimation.py [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+"""
+Title: An approach towards Depth Estimation with CNN.
+
+Author: [Victor Basu](https://www.linkedin.com/in/victor-basu-520958147)
+
+Date created: 2021/08/08
+
+Last modified: 2021/08/09
+
+Description: Implement an depth estimation model with CNN.
```
Comment (s-mrb) [inline examples/vision/depth_estimation.py:17]: Redundant

File: examples/vision/depth_estimation.py [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+
+**Depth Estimation** is a crucial step towards inferring scene geometry from 2D images.
+The goal in monocular Depth Estimation is to predict the depth value of each pixel, given
+only a single RGB image as input.
+
+This is an approach to build a depth estimation model with CNN and basic loss functions.
+While I was working on this topic I came across various research paper that explains
+solving this problem and to be honest my solution is not as good as in those papers. I
+would suggest you to go through those, just search depth estimation on
+"paperwithcode.com".
```
Comment (s-mrb) [inline examples/vision/depth_estimation.py:31]: This seems un-necessary, you can add it in references or further read section.

File: examples/vision/depth_estimation.py [hunk 6]
```diff
@@ -0,0 +1,507 @@
...
+
+"""
+
+"""
+##Setup
+"""
+
+import os
+import cv2
+import csv
```
Comment (s-mrb) [inline examples/vision/depth_estimation.py:44]: needless import.

File: examples/vision/depth_estimation.py [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+sess = tf.compat.v1.Session(graph=tf.compat.v1.get_default_graph(), config=session_conf)
+tf.compat.v1.keras.backend.set_session(sess)
+logging.disable(sys.maxsize)
+
+"""
+## Download the dataset
+
+We will be using the **DIODE: A Dense Indoor and Outdoor Depth Dataset**  for this
+tutorial. We have used the validation set for training and validating our model. The
+reason we have used validation set and not training set of the orginal dataset because
```
Comment (s-mrb) [inline examples/vision/depth_estimation.py:71]: typo `orginal`

File: examples/vision/depth_estimation.py [hunk 8]
```diff
@@ -0,0 +1,507 @@
...
+the training set consist of 81GB data which was a bit difficult to download compared to
```
Comment (s-mrb) [inline examples/vision/depth_estimation.py:72]: > consist`s` of 81GB `of` data

File: examples/vision/depth_estimation.py [hunk 9]
```diff
@@ -0,0 +1,507 @@
...
+### angle-3
+"""
+
+ax.view_init(45, 220)
+fig
+
+"""
+## Building the model
+1. The basic model architecture has been taken from U-NET.
+2. Residual-blocks has been used in the down-scale blocks of the U-NET architecture.
```
Comment (s-mrb) [inline examples/vision/depth_estimation.py:265]: > have been used

File: examples/vision/depth_estimation.py [hunk 10]
```diff
@@ -0,0 +1,507 @@
...
+for i in range(6):
+    ax[i, 0].imshow((input[i].squeeze()))
+    ax[i, 1].imshow((target[i].squeeze()), cmap=cmap)
+    ax[i, 2].imshow((pred[i].squeeze()), cmap=cmap)
+
+"""
+## Scopes of Improvement
+
+1. From the research papers that I read while I was working on this topic the encode part
+of the unet was replaced with DenseNet, ResNet or other pre-trained model, which could be
```
Comment (s-mrb) [inline examples/vision/depth_estimation.py:501]: > `U-Net` instead of `unet`

**Review by s-mrb — COMMENTED** [review]
initial review

File: examples/vision/depth_estimation.py (hunk 3 above)
Comment (s-mrb) [reply examples/vision/depth_estimation.py:6]: Please remove full stop from end of line

File: examples/vision/depth_estimation.py [hunk 11]
```diff
@@ -0,0 +1,507 @@
...
+
+import tensorflow as tf
+from tensorflow.keras import layers
+import tensorflow.keras.backend as K
+
+import pandas as pd
+import numpy as np
+import matplotlib.pyplot as plt
+from PIL import Image
+from sklearn.utils import shuffle
```
Comment (s-mrb) [inline examples/vision/depth_estimation.py:56]: you can do shuffling without sklearn.shuffle.

File: examples/vision/depth_estimation.py [hunk 12]
```diff
@@ -0,0 +1,507 @@
...
+visualize_samples = next(
+    iter(DataGenerator(data=df, batch_size=6, dim=(config.HEIGHT, config.WIDTH)))
+)
+input, target = visualize_samples
+cmap = plt.cm.jet
+cmap.set_bad(color="black")
+fig, ax = plt.subplots(6, 2, figsize=(50, 50))
+for i in range(6):
+    ax[i, 0].imshow((input[i].squeeze()))
+    ax[i, 1].imshow((target[i].squeeze()), cmap=cmap)
```
Comment (s-mrb) [inline examples/vision/depth_estimation.py:223]: Please update this code, it gonna make unpleasant display.
More-ever you are using it three times so it would be better if you make a reusable function for it.

**Review by s-mrb — COMMENTED** [review]
missed this.
gonna run code then will send further review :)

**Comment by victor369basu** [comment]
Thank you very much for your time and consideration in reviewing my code. I have updated the code as you suggested. Please do consider re-reviewing my code.

File: examples/vision/depth_estimation.py (hunk 1 above)
Comment (victor369basu) [reply examples/vision/depth_estimation.py]: updated

File: examples/vision/depth_estimation.py (hunk 3 above)
Comment (victor369basu) [reply examples/vision/depth_estimation.py]: updated

File: examples/vision/depth_estimation.py (hunk 7 above)
Comment (victor369basu) [reply examples/vision/depth_estimation.py]: updated

File: examples/vision/depth_estimation.py (hunk 9 above)
Comment (victor369basu) [reply examples/vision/depth_estimation.py]: updated

File: examples/vision/depth_estimation.py (hunk 12 above)
Comment (victor369basu) [reply examples/vision/depth_estimation.py]: updated

File: examples/vision/depth_estimation.py (hunk 11 above)
Comment (victor369basu) [reply examples/vision/depth_estimation.py]: updated

File: examples/vision/depth_estimation.py (hunk 10 above)
Comment (victor369basu) [reply examples/vision/depth_estimation.py]: updated

File: examples/vision/depth_estimation.py (hunk 9 above)
Comment (victor369basu) [reply examples/vision/depth_estimation.py]: updated

File: examples/vision/depth_estimation.py (hunk 8 above)
Comment (victor369basu) [reply examples/vision/depth_estimation.py]: updated

File: examples/vision/depth_estimation.py (hunk 4 above)
Comment (victor369basu) [reply examples/vision/depth_estimation.py]: updated

File: examples/vision/depth_estimation.py (hunk 2 above)
Comment (victor369basu) [reply examples/vision/depth_estimation.py]: updated

**Comment by victor369basu** [comment]
hello, @fchollet
I haven't received any response from the reviewer. if I have missed out on anything or have done any mistake in the pull request do let me know. Thank you.

File: examples/vision/depth_estimation.py [hunk 13]
```diff
@@ -0,0 +1,516 @@
...
+            dim=(config.HEIGHT, config.WIDTH),
+        )
+    )
+)
+visualizeDepthMap(test_loader, test=True)
+
+"""
+## Scopes of Improvement
+
+1. From the research papers that I read while I was working on this topic the encode part
```
Comment (mattdangerw) [inline examples/vision/depth_estimation.py:490]: I would rewrite these two sections to be a bit less conversational and more information dense. Prefer second person (E.g. You could improve this model by...)

**Comment by mattdangerw** [comment]
That's my bad! I started a review last week, but had wanted to dig a little deeper into the input pipeline to see if there are changes we could make there.

Let me post the comments I have now so you can start working. Sorry for the delay!

**Review by mattdangerw — CHANGES_REQUESTED** [review]

File: examples/vision/depth_estimation.py [hunk 14]
```diff
@@ -0,0 +1,516 @@
...
+We have tried to optimize 3 losses in our model.
+1. Structural similarity index(SSIM).
+2. L1-loss, or Point-wise depth in our case.
+3. Edge wide depth with depth smoothness.
+
+Out of the three loss functions SSIM contributed the most in improving model performance.
+"""
+
+
+class DepthEstimationModel(tf.keras.Model):
```
[the remark this answers is not in the document]
Comment (victor369basu) [reply examples/vision/depth_estimation.py:352]: In my recent commit/ updated code I have updated everything as you suggested except for this part. I found it a bit difficult to combine them into a single subclass model in my case, which maybe because of my lack of knowledge on that. if this creates an issue for approval could you please suggest to me an example so that I could understand it in a better way? i followed Keras examples like [image captioning](https://keras.io/examples/vision/image_captioning/) and [oxford_pets_image_segmentation](https://keras.io/examples/vision/oxford_pets_image_segmentation/)

File: examples/vision/depth_estimation.py [hunk 15]
```diff
@@ -0,0 +1,489 @@
...
+
+    def __getitem__(self, index):
+        if (index + 1) * self.batch_size > len(self.indices):
+            self.batch_size = len(self.indices) - index * self.batch_size
+        # Generate one batch of data
+        # Generate indices of the batch
+        index = self.indices[index * self.batch_size : (index + 1) * self.batch_size]
+        # Find list of IDs
+        batch = [self.indices[k] for k in index]
+        X, y = self.__data_generation(batch)
```
Comment (fchollet) [inline examples/vision/depth_estimation.py:134]: Style nit: always use lowercase for variable name

File: examples/vision/depth_estimation.py [hunk 16]
```diff
@@ -0,0 +1,489 @@
...
+        image_ = cv2.cvtColor(image_, cv2.COLOR_BGR2RGB)
+        image_ = cv2.resize(image_, self.dim)
+        image_ = tf.image.convert_image_dtype(image_, tf.float32)
+
+        depth_map = np.load(depth_map).squeeze()
+
+        mask = np.load(mask)
+        mask = mask > 0
+
+        MIN_DEPTH = 0.5
```
Comment (fchollet) [inline examples/vision/depth_estimation.py:160]: This seems like it should be a (lowercase) instance attribute defined in `__init__()`

File: examples/vision/depth_estimation.py [hunk 17]
```diff
@@ -0,0 +1,489 @@
...
+
+        MAX_DEPTH = min(300, np.percentile(depth_map, 99))
```
Comment (fchollet) [inline examples/vision/depth_estimation.py:162]: Same

File: examples/vision/depth_estimation.py [hunk 18]
```diff
@@ -0,0 +1,489 @@
...
+    def on_epoch_end(self):
+
+        """
+        Updates indexes after each epoch
+        """
+        self.index = np.arange(len(self.indices))
+        if self.shuffle == True:
+            np.random.shuffle(self.index)
+
+    def __load__(self, image_path, depth_map, mask):
```
Comment (fchollet) [inline examples/vision/depth_estimation.py:147]: Doesn't need to be private

File: examples/vision/depth_estimation.py [hunk 19]
```diff
@@ -0,0 +1,489 @@
...
+            [y] * 3,
+            [x] * 3,
+            c=tuple(img_vis[x, y, :3] / 255),
+            s=3,
+        )
+    ax.view_init(45, 135)
+
+"""
+## Building the model
+1. The basic model architecture have been from U-Net.
```
Comment (fchollet) [inline examples/vision/depth_estimation.py:244]: Add line break above

File: examples/vision/depth_estimation.py [hunk 20]
```diff
@@ -0,0 +1,489 @@
...
+"""
+## Model training
+"""
+
+optimizer = tf.keras.optimizers.Adam(
+    learning_rate=LR,
+    amsgrad=False,
+)
+model = unet_model(HEIGHT, WIDTH)
+DEM = DepthEstimationModel(model=model)
```
Comment (fchollet) [inline examples/vision/depth_estimation.py:427]: Lowercase

**Review by fchollet — COMMENTED** [review]
Just a few style nits.

File: examples/vision/depth_estimation.py [hunk 21]
```diff
@@ -0,0 +1,489 @@
...
+DEM.compile(optimizer)
+
+train_loader = DataGenerator(
+    data=df[:600].reset_index(drop="true"), batch_size=BATCH_SIZE, dim=(HEIGHT, WIDTH)
+)
+DEM.fit(train_loader, epochs=EPOCHS)
```
Comment (fchollet) [inline examples/vision/depth_estimation.py:433]: Please monitor performance on a validation set so you can avoid overfitting and have some visibility into generalization power

File: examples/vision/depth_estimation.py [hunk 22]
```diff
@@ -0,0 +1,512 @@
...
+            c=tuple(img_vis[x, y, :3] / 255),
+            s=3,
+        )
+    ax.view_init(45, 135)
+
+"""
+## Building the model
+
+1. The basic model architecture have been from U-Net.
+2. Residual-blocks has been used in the down-scale blocks of the U-Net architecture.
```
[the remark this answers is not in the document]
Comment (victor369basu) [reply examples/vision/depth_estimation.py]: I am replacing this line with "Skip connection via addition is implemented in the down-scale block." as I have used skip connection in the downscale block

File: examples/vision/depth_estimation.py [hunk 23]
```diff
@@ -0,0 +1,488 @@
...
+## Building the model
+
+1. The basic model is a U-Net.
+2. Addditive skip-connections are implemented in the downscaling block.
+"""
+
+
+class DownscaleBlock(layers.Layer):
+    def __init__(self, filters, kernel_size=(3, 3), padding="same", strides=1):
+        super().__init__()
```
Comment (fchollet) [inline examples/vision/depth_estimation.py:240]: In all custom layers you define, add `**kwargs` in the signature and pass them to `super().__init__(**kwargs)`

File: examples/vision/depth_estimation.py [hunk 24]
```diff
@@ -0,0 +1,488 @@
...
+        dy_true, dx_true = tf.image.image_gradients(target)
+        dy_pred, dx_pred = tf.image.image_gradients(pred)
+        weights_x = tf.exp(tf.reduce_mean(tf.abs(dx_true)))
+        weights_y = tf.exp(tf.reduce_mean(tf.abs(dy_true)))
+
+        # Depth smoothness
+        smoothness_x = dx_pred * weights_x
+        smoothness_y = dy_pred * weights_y
+
+        l_edges = tf.reduce_mean(abs(smoothness_x)) + tf.reduce_mean(abs(smoothness_y))
```
Comment (fchollet) [inline examples/vision/depth_estimation.py:351]: Use a full name, not "l"

File: examples/vision/depth_estimation.py [hunk 25]
```diff
@@ -0,0 +1,488 @@
...
+
+"""
+## Model training
+"""
+
+optimizer = tf.keras.optimizers.Adam(
+    learning_rate=LR,
+    amsgrad=False,
+)
+dem = DepthEstimationModel()
```
Comment (fchollet) [inline examples/vision/depth_estimation.py:421]: Just "model"

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the updates, @victor369basu !

And thanks for the review, @mattdangerw !

I pushed some copyedits and made some last comments.

**Comment by fchollet** [comment]
Please reapply the changes from https://github.com/keras-team/keras-io/pull/584/commits/8cbdec2fac393cc0b70c028c95de9204d633fe4e

**Comment by victor369basu** [comment]
@fchollet I have re-applied changes from 8cbdec2.

**Review by fchollet — APPROVED** [review]
Thanks for the update! I believe we're ready to publish. Please add the generated files.

**Comment by victor369basu** [comment]
@fchollet I have added the generated files.

**Comment by fchollet** [comment]
Thanks for the great contribution! And thanks again @mattdangerw for the review.
